cmd/kube-proxy/app/server.go

```diff
@@ -263,7 +263,7 @@ func (o *Options) eventHandler(ent fsnotify.Event) {
 	eventOpIs := func(Op fsnotify.Op) bool {
 		return ent.Op&Op == Op
 	}
-	if eventOpIs(fsnotify.Write) || eventOpIs(fsnotify.Rename) {
+	if eventOpIs(fsnotify.Remove) || eventOpIs(fsnotify.Write) || eventOpIs(fsnotify.Rename) {
```

Comment:
I wouldn't expect us to restart when the config file is removed... that will prevent it from starting back up, right?

Do we need to watch the parent dir to notice when a file is recreate?

Reply:
that's a good question, should kube-proxy work without config file then?
maybe if the file is removed kube-proxy should stop working

Reply:
I didn't think the goal was to make kube-proxy stop working if the config file is removed, I thought it was to make kube-proxy restart when a new valid config file is present

I expected a change in the config file to trigger validation of the file, and if valid, to restart. Looking at the current code, it appears that any change or removal immediately restarts kube-proxy, which means an invalid config will make all kube-proxy instances immediately become unavailable?

Reply:
Thanks @liggitt I agree with you. But if I modified the kube-proxy configmap, kube-proxy only can receive `CHMOD` and `REMOVE` events. Although the config file is overwritten/created, no other events are detected. Is this a bug with configmap or kubelet?

Reply:
I think it's a bug in the kube-proxy, but I'm not sure. I think we need to watch fs events on the parent dir to detect the creation of a new file at the specified location.

Once we detect a new file, I would expect us to validate it before exiting.

Reply:
then .... are not `remove` and `rename` the same type of event here?
the result for kube-proxy is the same, right? both remove the current configuration file and restart without checking the new config.

Reply:
> are not `remove` and `rename` the same type of event here?

I'm not very familiar with the fs event handling (would recommend asking someone in sig-node or sig-storage), but I vaguely remember having to watch the parent dir and filter to the desired path to be sure all delete/recreate scenarios were observed.

> restart without checking the new config

yes, that seems like something we should fix

Reply:
This can get tricky if it doesn't restart directly. 
In the mode of mounting kube-proxy configmap, the config file is a soft link to a real config file (folder is named by time).
```
drwxrwxrwx 3 root root 101 Jan  9 03:52 .
drwxr-xr-x 1 root root  24 Jan  9 03:44 ..
drwxr-xr-x 2 root root  48 Jan  9 03:52 ..2020_01_09_03_52_40.739411708
lrwxrwxrwx 1 root root  31 Jan  9 03:52 ..data -> ..2020_01_09_03_52_40.739411708
lrwxrwxrwx 1 root root  18 Jan  9 03:44 config.conf -> ..data/config.conf
lrwxrwxrwx 1 root root  22 Jan  9 03:44 kubeconfig.conf -> ..data/kubeconfig.conf
```
The config file is updated by creating new config file and modifying the target of the soft link, but the soft link itself never changes(we can not receive any fs event about it).
If we watch the parent directory, we can only observe these events:
```
W0109 03:46:02.646560       1 server.go:272] Found Configuration Change event "/var/lib/kube-proxy/..2020_01_09_03_46_02.974746577": CREATE
W0109 03:46:02.646589       1 server.go:272] Found Configuration Change event "/var/lib/kube-proxy/..2020_01_09_03_46_02.974746577": CHMOD
W0109 03:46:02.646660       1 server.go:272] Found Configuration Change event "/var/lib/kube-proxy/..data_tmp": RENAME
W0109 03:46:02.646683       1 server.go:272] Found Configuration Change event "/var/lib/kube-proxy/..data": CREATE
W0109 03:46:02.646741       1 server.go:272] Found Configuration Change event "/var/lib/kube-proxy/..2020_01_09_03_44_54.771236680": REMOVE
```
As you can see above, we can not simply judge file changes by matching path(`/var/lib/kube-proxy/config.conf`).
We need to deal with complex link relationships to determine whether the config file has been modified or not. In special cases, watching the parent directory might receive too many unrelated events( e.g. config file in the root directory).

We can avoid this complexity if we choose to simply restart kube-proxy.
Looking forward to your Suggestions @liggitt @aojea

Reply:
I think there are 2 different things here and it depends if we want to modify current behavior.

Reading the code and the PR that introduced this feature #59176 it seems that the expected behavior is to restart if the configuration file changes, no matter what, because the configuration file is generated when kube-proxy `Runs()`

https://github.com/kubernetes/kubernetes/blob/95debf1a3a746004b7f692ec150a854e69c2ebb7/cmd/kube-proxy/app/server.go#L307-L309

If we want to keep current behavior I think that this change is fine because I don't see any difference, from the kube-proxy perspective, between `delete` or `rename`, i.e.:
if kube-proxy uses `..data/config.conf` , doesn't matter if you remove the file or rename it to `..data/config.conf.bak`, it won't use that file and will generate a new one from the configmap.
My 2 cents :-)

Reply:
> In special cases, watching the parent directory might receive too many unrelated events( e.g. config file in the root directory).

Yes, events received by watching the parent would have to be filtered, or we would need to do something different like poll config file content periodically. For what it's worth, other components like the kube-apiserver poll periodically to support dynamic reloading of serving certificates (poll periodically, and if content has changed and validates, take action to replace... for kube-proxy, that would mean restarting).

> We can avoid this complexity if we choose to simply restart kube-proxy.

Complexity that prevents a faulty or missing config from taking down all kube-proxy instances might be worthwhile.

> I think there are 2 different things here and it depends if we want to modify current behavior.

I'm not seeing the direct benefit of restarting the kube-proxy process when its configuration file is *removed* (it assumes a removal is the first step of a replacement, but we don't actually know that). If the goal is to make the kube-proxy process restart when a different config file is placed at that path, then we should implement that instead.

> the configuration file is generated when kube-proxy Runs()

Can you clarify what you mean by that? The kube-proxy does not write to the configuration file when run normally that I'm aware of.

Reply:
> > the configuration file is generated when kube-proxy Runs()
> 
> Can you clarify what you mean by that? The kube-proxy does not write to the configuration file when run normally that I'm aware of.

scratch it, sorry,  I mixed things here,  the `o.WriteConfigTo` is other thing
	> fs.StringVar(&o.WriteConfigTo, "write-config-to", o.WriteConfigTo, "If set, write the default configuration values to this file and exit.")